### ISSUE TITLE
Tracking Issue for `ptr::Alignment` type

### ISSUE THREAD

**Issue by scottmcm** [rust-lang/rust#102070]
Feature gate: `#![feature(ptr_alignment)]`

This is a tracking issue for the `ptr::Alignment` type, and related methods, to represent values that are valid alignments in the 

### Public API

```rust
// core::ptr

pub struct Alignment(…);

impl Magic {
    // TBD
}
```

### Steps / History

- [x] ACP https://github.com/rust-lang/libs-team/issues/108
- [ ] Implementation: #...
- [ ] Final comment period (FCP)[^1]
- [ ] Stabilization PR

### Unresolved Questions

- None yet.

[^1]: https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html

**Comment by kupiakos** [comment]
I've been able to take advantage of an alignment newtype for performance in an embedded arena library and would find this useful to be stabilized. Are there any blockers for an FCP?

**Comment by m-ou-se** [comment]
We discussed this in the libs-api meeting just now. We think this type should be used in std APIs (e.g. methods of Layout) before stabilizing the type. Right now, this type doesn't seem to be used anywhere (publicly) in std, so we'd like to see a plan for usage of this type in std.

We also briefly talked about the error type in the TryFrom impls. It's not immediately clear if TryFromIntError is the right type. Perhaps it needs a new error type.

**Comment by clarfonthey** [comment]
Genuine question, as I'm looking at this API: are alignments above `isize::MAX` actually allowed, given how most allocations are limited to `isize::MAX` instead of `usize::MAX`?

**Comment by scottmcm** [comment]
1. Yes, because you can make `isize::MAX+1`-aligned ZST at the address `ptr::invalid(isize::MIN as usize)`.  That's probably not all that *useful*, but it's legal in the rust AM, as far as I know.
5. Agreed, having good ways to use this for things like masking (or aligning up/down) are important.  I haven't gotten to making specifics for that; I'd encourage you to make some if  you have something in mind.

**Comment by Gnurou** [comment]
Hi,

This work has been brought to my attention as I've been working on a [very similar thing](https://lore.kernel.org/rust-for-linux/20250620-num-v1-0-7ec3d3fb06c9@nvidia.com/T/#m8ed9d14f65e5294ba00f2f499361e3c028bac66b) in Rust for Linux, where we regularly need to align values up or down to a power of two. Having discovered this, I would like to align my implementation and hopefully replace it with this feature when it lands. A few questions though:

- Are there blockers to merge this? This look pretty close to what we need in R4L.
- We are primarily using the `align_up` and `align_down` methods of our type in R4L. These methods don't exist yet in `Alignment`. Can we hope to get them? I'd argue that they are a good complement to `mask`.
- Does this type need to be limited to the host's pointer size? R4L has use-cases for this to e.g. compute alignments for a microcontroller, where the pointer size is not necessarily the same as the CPU's. We addressed this by making our type generic against any unsigned.

**Comment by clarfonthey** [comment]
The type is only specific to alignment in the sense that the maximum alignment will be `isize::MAX` on the host, which is going to be far larger than any possible alignment you could want anyway. (On 16-bit hosts, 32KiB is *almost* reasonable, but 2GiB for 32-bit hosts, and 8EiB for 64-bit hosts, is far beyond what you'd want.)

`align_up` and `align_down` are equivalent to bit-shifting, and while I don't personally see an issue with them, it would be worth asking what the exact overflow semantics should be for them. Should `align_down` simply be saturating, or checked? Should `align_up` be checked or strict? Is it okay to have asymmetric behaviour for these two?

I don't think there are any blockers as much as this API hasn't been really explored much; you're the first ones to discuss seriously using it here.

**Comment by Gnurou** [comment]
Thanks for the comment! You are right that too large alignments won't make much sense, although for safety I wish we could go up to 4GB, just in case. 😅

When aligning for e.g. a 16-bit address space we would need to do some fallible type conversions, but that's probably something we can live with (and worst case, we can define a new type on top of `Alignment`).

Regarding `align_down`, I don't think there is a need for a checked version, as `0` is always a valid minimal aligned value (unless I missed something). `align_up` should probably be checked though.

I am tempted to propose enabling the feature in the kernel so this code gets some exercise, as this sounds like it could help with merging it?

**Comment by ojeda** [comment]
> I am tempted to propose enabling the feature in the kernel so this code gets some exercise, as this sounds like it could help with merging it?

It mostly depends if the feature is expected to become stable or not or, at least, if Rust thinks it will not change often until stabilization.

**Comment by clarfonthey** [comment]
> Thanks for the comment! You are right that too large alignments won't make much sense, although for safety I wish we could go up to 4GB, just in case. 😅

I mean, if you're running the code on a 64-bit host, that isn't a problem; since you're computing alignments *for* a microcontroller instead of *from* a microcontroller, it's probably fine. Ultimately, this type is going to probably remain specifically for Rust alignments, so, you're going to run into trouble if you want to align to 4 GiB on a 32-bit host, for example.

That said, the desire for other use cases probably makes adding an auxiliary type useful.

<details><summary>(irrelevant; misreading that 0 is an invalid aligned value, thinking they meant alignment)</summary>

> Regarding `align_down`, I don't think there is a need for a checked version, as `0` is always a valid minimal aligned value (unless I missed something). `align_up` should probably be checked though.

Zero is explicitly *not* a valid alignment: the minimum alignment is one, since you can't divide by zero. And to clarify, by "checked", I mean whether you want to explicitly return `None` on `align_down` from 1, or just 1 again. Aligning up, it's unclear whether you'd want to always have to check if your alignment is `None`, or to just panic if you overflow. I think we'd always explicitly check anyway, since overflowing to zero would be UB. (as mentioned before)

---

Worth mentioning also that I mentioned earlier adding a `DoubledLen` type (name pending) which is just a power-of-two, allowing zero. Not sure if zero is specifically useful for your use case or if you just misremembered something.

</details>

**Comment by LGFae** [comment]
> Zero is explicitly not a valid alignment: the minimum alignment is one, since you can't divide by zero.

Sorry to butt in, but I don't think that is what @Gnurou meant. I think he was talking that the 0 address is aligned to any value, since 0 is always divisible by any number different than 0. If you want to program microcontrollers' kernels, the 0 address could indeed be valid, and will always be aligned.

**Comment by clarfonthey** [comment]
You're absolutely right to butt in; I misread the comment and you're right that the 0 address is always aligned.

**Comment by hanna-kruppe** [comment]
Alignment limits vary enormously by context. For things that the linker has to align (e.g., [statics](https://github.com/rust-lang/rust/issues/70022) and [functions](https://github.com/rust-lang/rust/issues/143368)), the limit is platform-dependent and often as small as a page or two. On the other hand, when managing virtual memory with huge pages, you can encounter alignments of 1 GiB or more. The library type `ptr::Alignment` might be useful in those cases even if the language had smaller limits for the alignments that language constructs deal in.

**Comment by scottmcm** [comment]
@Gnurou Some kind of methods like that make a ton of sense to me, but as I was starting to think about it I was unsure exactly the best way to spell them.

If I'm reading that diff right, it looks like they're written to just take `usize`?  I kinda feel like in Rust I'd expect to have them on the pointer types directly so I don't need to use `map_addr` with it.  On nightly today you can write `align_down` as `ptr.mask(alignment.mask())`, for example.  But there's also the big question when working on pointers of whether these use `wrapping_offset` or `offset` under the hood, and thus whether they have the "stay within the allocation" restriction on them.

So I'd encourage you to start [a new ACP](https://github.com/rust-lang/libs-team/issues/new?template=api-change-proposal.md) with your vision for how you'd like those methods to work, so the conversation can happen there.  I can see a couple of different kinds of solutions:
1. Maybe you're looking for more like a `num::PowerOfTwo<T>` type if you want to specifically work on integers only, and need to use things *other* than just `usize`.
2. Maybe you want *unsafe* methods for `align_up`/`align_down` on `*const T`/`*mut T`/`NotNull<T>` that go with [`is_aligned_to`](https://doc.rust-lang.org/nightly/std/primitive.pointer.html#method.is_aligned_to), changing that one to also take `ptr::Alignment`, which use `offset` under the hood and don't need to worry about overflows because that's impossible inside an allocation
3. Maybe you want something like the previous but as `wrapping_round_up`/`wrapping_round_down` that can be safe but wrapping up might give a null pointer so it can't be on `NotNull<T>`.
4. Maybe you want something like those but in `checked_` versions that return `Option<Self>`s in some situations.
5. Maybe you want `round_up`/`round_down` on integers taking a `NonZero<T>` as the argument, so you can use that with [`map_addr`](https://doc.rust-lang.org/nightly/std/primitive.pointer.html#method.map_addr) to adjust things

**Comment by joboet** [comment]
Is there a reason that `Alignment` lives in `ptr` and not `mem`? Given that `size_of` and `align_of` are both located in `mem` that seems a bit weird to me...

**Comment by scottmcm** [comment]
This was added [way back](https://github.com/rust-lang/libs-team/issues/108#issuecomment-1252326510) when it just needed a second, so there might have never been a team discussion of where it should live.

@clarfonthey I don't know of anything, personally.  Maybe made a partial stabilization PR and ask libs-api to weigh in?  (I suggest the PR because for anything partial I think it's easier to understand *exactly* what's being stabilized that way.)  Seems entirely plausible to start with just "well we have the type and the obvious conversion methods" and the various helpers and other stuff can be discussed later.  (Dunno if it'd be worth including <https://doc.rust-lang.org/nightly/std/alloc/struct.Layout.html#method.alignment> in the first pass.)

**Comment by scottmcm** [comment]
`as_isize` doesn't make sense to me because it would have to either be fallible or give `isize::MIN` sometimes.

**Comment by scottmcm** [comment]
@clarfonthey Which now-merged PR are you asking about?  Can you link the line in the PR diff?

Oh, do you mean  https://github.com/rust-lang/rust/pull/154004/changes#diff-7f6a8c7b45272fd1389c2de21aadb62686bb7d15e4cf2d82a27eff37757169bfL183-R183  ?

That's because AFAIK *macros* can't use lang items and need to generate the path that the compiler will then parse and resolve.  The compiler will use the lang item if it just needs a `DefId` internally for something (like generating the shims) but when the proc macro (even the magic internal proc macro) references something it's by path.  (Or, at least, that's how all the derives work when they need to mention things, last I looked.)

**Comment by herabit** [comment]
I would like to add that if there's an `Alignment` type, would it not additionally make sense to add a `Mask` type (whatever you dictate that to be) and `Size` type?

These types would benefit from niche layout optimizations, as well as providing optimization hints, such as using the assume intrinsic when accessing the values of `Alignment` that they're always powers-of-two.

**Comment by clarfonthey** [comment]
I agree that a `Size` intrinsic *would* be nice, although it's more likely best to wait until some flavour of generic integers or `NonNegative<isize>` instead of that, whereas `Alignment` is explicitly `PositivePowerOfTwo<isize>` which is less likely to be generally applicable outside alignments.

That said, I *have* desired some form of `Option<Alignment>` that operates as a len-niche-type, and it appears that `RawVec` already uses something similar to it.

**Comment by bjoernager** [comment]
Can we rename the method to `as_non_zero_usize` now? With `Box::from_non_null` et al. stabilised in #157226.
